src/main/java/com/iota/iri/model/Hash.java

```diff
@@ -10,85 +10,125 @@
+        byte[] dest = new byte[SIZE_IN_BYTES];
+        // TODO - THIS IS A BUGGY BOUNDS-SIZE CHECK. I'LL KEEP IT IN BECAUSE TOO MANY TESTS FAIL OTHERWISE.
+        System.arraycopy(source, sourceOffset, dest, 0, sourceSize - sourceOffset > source.length ? source.length - sourceOffset : sourceSize);
+        this.byteSafe = new ByteSafe(dest);
     }
 
-    //
-    /*
-    public static Hash calculate(byte[] bytes) {
-        return calculate(bytes, SIZE_IN_TRITS, new Curl());
+    public Hash(final byte[] bytes) {
+        this(bytes, 0, SIZE_IN_BYTES);
     }
-    */
-    public static Hash calculate(SpongeFactory.Mode mode, int[] trits) {
-        return calculate(trits, 0, trits.length, SpongeFactory.create(mode));
+
+    public Hash(final int[] trits, final int offset) {
+        int[] dest = new int[SIZE_IN_TRITS];
+        System.arraycopy(trits, offset, dest, 0, dest.length);
+        this.tritSafe = new TritSafe(dest);
     }
 
-    public static Hash calculate(byte[] bytes, int tritsLength, final Sponge curl) {
-        int[] trits = new int[tritsLength];
-        Converter.getTrits(bytes, trits);
-        return calculate(trits, 0, tritsLength, curl);
+    public Hash(final int[] trits) {
+        this(trits, 0);
     }
-    public static Hash calculate(final int[] tritsToCalculate, int offset, int length, final Sponge curl) {
-        int[] hashTrits = new int[SIZE_IN_TRITS];
-        curl.reset();
-        curl.absorb(tritsToCalculate, offset, length);
-        curl.squeeze(hashTrits, 0, SIZE_IN_TRITS);
-        return new Hash(hashTrits);
+
+    public Hash(final String trytes) {
+        this.tritSafe = new TritSafe(new int[SIZE_IN_TRITS]);
+        Converter.trits(trytes, this.tritSafe.trits, 0);
+    }
+
+    private void fullRead(byte[] src) {
+        if (src != null) {
+            synchronized (lock) {
+                if (byteSafe != null || tritSafe != null) {
+                    throw new IllegalStateException("I cannot be initialized with data twice.");
+                }
+                byte[] dest = new byte[SIZE_IN_BYTES];
+                System.arraycopy(src, 0, dest, 0, Math.min(dest.length, src.length));
+                byteSafe = new ByteSafe(dest);
+            }
+        }
     }
 
     public int trailingZeros() {
-        int index, zeros;
-        final int[] trits;
-        index = SIZE_IN_TRITS;
-        zeros = 0;
-        trits = trits();
-        while(index-- > 0 && trits[index] == 0) {
+        final int[] trits = trits();
+        int index = SIZE_IN_TRITS;
+        int zeros = 0;
+        while (index-- > 0 && trits[index] == 0) {
             zeros++;
         }
         return zeros;
     }
 
     public int[] trits() {
-        if(trits == null) {
-            trits = new int[Curl.HASH_LENGTH];
-            Converter.getTrits(bytes, trits);
+        TritSafe safe = tritSafe;
+        if (safe == null) {
+            synchronized (lock) {
```

Comment:
Good implementation double-checked locking!
Here is a link for others who may inspecting the code:
https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java